nixos/modules/services/hardware/udev.nix

```diff
@@ -418,18 +418,9 @@ in
     # We don't place this into `extraModprobeConfig` so that stage-1 ramdisk doesn't bloat.
     environment.etc."modprobe.d/firmware.conf".text = "options firmware_class path=${config.hardware.firmware}/lib/firmware";
 
-    system.activationScripts.udevd =
-      ''
-        # The deprecated hotplug uevent helper is not used anymore
-        if [ -e /proc/sys/kernel/hotplug ]; then
-          echo "" > /proc/sys/kernel/hotplug
-        fi
-
-        # Allow the kernel to find our firmware.
-        if [ -e /sys/module/firmware_class/parameters/path ]; then
-          echo -n "${config.hardware.firmware}/lib/firmware" > /sys/module/firmware_class/parameters/path
-        fi
-      '';
+    systemd.tmpfiles.rules = [
+      "w /sys/module/firmware_class/parameters/path - - - - '${config.hardware.firmware}/lib/firmware'"
```

Comment:
`systemd-tmpfiles-setup.service` is probably far too late during boot for this. Plenty of devices are found and used arbitrarily early, and they could very well need firmware.

Reply:
Should we throw this into preStart for `systemd-udev` then?

Reply:
Firmware won't be loaded until the corresponding module is loaded, and the module won't be loaded without udev (or something else in userspace) causing that, so I think that should be fine.

Reply:
@lheckemann I was under the impression that the kernel itself would use the contents of `/proc/sys/kernel/modprobe` to load modules when devices requiring them are found. Is that false? Because that would mean that both this and the `systemd-modules-load.preStart` have to happen before stage 2 starts. I could be totally wrong though; maybe it's *only* userspace that causes modules to be loaded?

Reply:
Even in the activationScripts no-one guarantees that the kernel might not request modules before the correct script is run. Why does this need to be done before systemd starts?

Reply:
First of all, `systemd.tmpfiles.rules` is not just after systemd starts; it's *way* after. There are plenty of devices that might need to appear before `local-fs.target`, which are automatically ruled out by having firmware loaded later than that because of `tmpfiles`.

Reply:
So should we order it before `local-fs-pre.target[¶](https://www.freedesktop.org/software/systemd/man/latest/systemd.special.html#local-fs-pre.target)`?

Again I'm very happy to run these as early as you want me  to. I don't have super strong feelings but I need the information when exactly to run them.

Reply:
I'm also unsure, unfortunately. All I can say is that there's very good reason that this happens before systemd becomes PID 1 currently.